### PR TITLE
jax lec for discrete optimal savings #10

### PR DESCRIPTION
**Pull request by chappiewuzefan**
This PR addresses #10

### CODE REVIEW

**Comment by chappiewuzefan** [comment]
I need double check below:
1. details for the lec
2. try run on GPU

**Comment by jstac** [comment]
Thanks @chappiewuzefan , please let me know when this is ready for review.

**Comment by jstac** [comment]
@chappiewuzefan Can you please confirm that this is ready for review?

In general, please add the 'ready' label and ping me when you want me to review.

@Smit-create , since you treated the investment jax lecture, which is very similar, can you please review this quickly after @chappiewuzefan has agreed it's ready?

File: lectures/opt_savings.md [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+
+```{code-cell} ipython3
+!pip install quantecon
+```
+
+We will use the following imports:
+
+```{code-cell} ipython3
+import numpy as np
+import quantecon as qe
```
Comment (Smit-create) [inline lectures/opt_savings.md:26]: Can you please try to remove `np.*` from this lecture - and use `jnp.*` instead?

File: lectures/opt_savings.md [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+                             γ=2.5,                     # CRRA parameter
+                             w_min=0.01,                # Min wealth
+                             w_max=5.0,                 # Max wealth
+                             w_size=150,                # Grid side
+                             ρ=0.9, ν=0.1, y_size=100): # Income parameters
+    """
+    A function that takes in parameters and returns an instance of Model that
+    contains data for the optimal savings problem.
+    """
+    w_grid = np.linspace(w_min, w_max, w_size)
```
Comment (Smit-create) [inline lectures/opt_savings.md:107]:
```suggestion
    w_grid = jnp.linspace(w_min, w_max, w_size)  
```

File: lectures/opt_savings.md [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+    yp_idx = jnp.arange(y_size)
+    yp_idx = jnp.reshape(yp_idx, (1, 1, y_size))
+    σ = jnp.reshape(σ, (w_size, y_size, 1))
+    V = v[σ, yp_idx]      
+
+    # Convert Q[j, jp] to Q[i, j, jp] 
+    Q = jnp.reshape(Q, (1, y_size, y_size))
+
+    # Calculate the expected sum Σ_jp v[σ[i, j], jp] * Q[i, j, jp]
+    Ev = np.sum(V * Q, axis=2)
```
Comment (Smit-create) [inline lectures/opt_savings.md:214]:
```suggestion
    Ev = jnp.sum(V * Q, axis=2)
```
... and wherever it applies.

File: lectures/opt_savings.md [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+σ_vfi, vfi_time = run_algorithm(value_iteration, model, tol=1e-5)
+
+m_vals = range(5, 3000, 100)
+opi_times = []
+for m in m_vals:
+    σ_opi, opi_time = run_algorithm(optimistic_policy_iteration, model, m=m, tol=1e-5)
+    opi_times.append(opi_time)
+
+fig, ax = plt.subplots(figsize=(9, 5.2))
+ax.plot(m_vals, np.full(len(m_vals), pi_time), lw=2, label="Howard policy iteration")
```
Comment (Smit-create) [inline lectures/opt_savings.md:448]: You can either use `jnp.full` or `[pi_time]* m_vals`. I don't think that would make much performance difference, but yeah that removes `numpy` completely.

**Review by Smit-create — COMMENTED** [review]
Thanks, @jstac and @chappiewuzefan. The lecture looks good to me. I have left some comments on removing `numpy`.

**Comment by chappiewuzefan** [comment]
@jstac ready for review

**Comment by Smit-create** [comment]
Thanks @chappiewuzefan. Looks good to me. Can you please remove the .ipynb file and checkpoints folder?

**Comment by chappiewuzefan** [comment]
@Smit-create thanks for reminding , done

**Comment by jstac** [comment]
Thanks @chappiewuzefan , and @Smit-create for the review.  I'll take it from here.
